debug: fallback to mark breakpoints verified on stopped/breakpoint events

amanmaurya92 opened the pull request:
## Description
LLDB-DAP sometimes fails to mark breakpoints as verified in the `setBreakpoints` response, even though the breakpoint is hit during debugging. This leads to the VS Code UI showing the breakpoint as “unverified” while execution actually stops there.

This PR adds a safe fallback that marks breakpoints as verified when we detect them being hit via:
- a `stopped` event with `reason: "breakpoint"`, or
- a `breakpoint` event containing a valid source and line.

This improves consistency between actual debugger behavior and the UI, especially for Swift executables and `swift test`.

Issue: https://github.com/swiftlang/vscode-swift/issues/1871

## Tasks
- [ ] Required tests have been written
- ~[ ] Documentation has been updated~
- [x] Added an entry to CHANGELOG.md if applicable

amanmaurya92 committed:
debug: fallback to mark breakpoints verified on stopped/breakpoint events

award999 reviewed (commented):
@amanthatdoescares thanks so much for the PR 🙏 I have some initial comments here to get you started. @matthewbastien is more familiar with the debugging codebase so let's have him take a look to.

Also the open todo for updating changelog in the PR template, you just need to do something like https://github.com/swiftlang/vscode-swift/pull/1979/commits/85d5132ed277fde61ec883ebfbf177a2310b155b, but this being a bug fix, you'd put it under the `### Fixed` section for the latest release.

amanmaurya92 committed:
Apply suggestions from code review

Co-authored-by: award999 <award999@apple.com>

amanmaurya92 committed:
Use location-based breakpoint filtering, add logger, normalize paths

amanmaurya92 committed:
Changelog: normalize adapter source.path for breakpoint fallback

amanmaurya92 committed:
Remove duplicate vscode-tmgrammar-test dependency

amanmaurya92 committed:
Remove duplicate license entry in package-lock.json

matthewbastien commented on src/debugger/logTracker.ts:
```diff
@@ -131,6 +136,126 @@ export class LoggingDebugAdapterTracker implements vscode.DebugAdapterTracker {
+
+                for (const bp of breakpoints) {
+                    const loc = bp.location;
+                    if (!loc) {
```
**suggestion:** This check is unnecessary as it should be handled by the filter above.

amanthatdoescares replied:
Good point, that check is redundant given the filter above. I’ve removed it.

matthewbastien reviewed (changes requested):
Thank you for fixing this! I just have some minor comments to make the code easier to read.

amanmaurya92 committed:
Update src/debugger/logTracker.ts

Co-authored-by: Matthew Bastien <matthew_bastien@apple.com>

amanmaurya92 committed:
Refine breakpoint fallback handling

amanmaurya92 committed:
Update changelog

amanmaurya92 committed:
Update CHANGELOG to remove version 2.14.2

Removed version 2.14.2 entry and its fixed items from the changelog.

amanmaurya92 committed:
Update CHANGELOG with new features and fixes

Added a note about using the JSON schema for editing config.json and fixed breakpoint filtering in debugger logging.

amanmaurya92 committed:
Merge branch 'main' into fix/breakpoint-verified-fallback

matthewbastien commented on CHANGELOG.md:
```diff
@@ -4,6 +4,10 @@
 
 ### Fixed
 
+- Fix breakpoint filtering fallback in debugger logging ([#1989](https://github.com/swiftlang/vscode-swift/pull/1989))
```
Thanks! The location looks good now.

Though I just noticed that the wording here would only be meaningful to contributors and probably won't make much sense to users of the extension. Maybe we could go with something like `Fixed breakpoints showing up as unresolved when using lldb-dap`?

amanthatdoescares commented:
@matthewbastien Thanks for the detailed review. After looking at it again, I realized I completely missed the real issue. My fallback logic only handles the happy case where the DAP breakpoint includes a source.path, and I didn’t actually test the scenario you ran into — the one where the adapter sends a breakpoint without a path. So the code ends up ignoring exactly the situation it was supposed to fix. Definitely a blooper on my part.

It’s clear now that this isn’t an environment/version difference — the logic just isn’t handling the missing-path case at all.
I’ll take another pass at this with the right approach:

- Reproduce the behavior properly and capture the actual setBreakpoints request + the breakpoint event.
- Fix the resolution logic so it works even when the adapter doesn’t provide a source.path.
- Add the integration test you mentioned: start a debug session with a breakpoint and assert that it gets verified once hit.

I’ll update the PR once I’ve got the full flow working as intended. Thanks again for pointing out.

matthewbastien committed:
pin the swiftlang/github-workflows configurations to 0.0.2 (#1993)

matthewbastien committed:
only attempt to run xcode-select on macOS (#1995)

matthewbastien committed:
Always use xcrun as a fallback for finding binaries on macOS (#1998)

award999 committed:
Ignore new nightly jobs (#1999)

plemarquand committed:
Revert "Add dynamic local schema detection for SourceKit-LSP (#1979)" (#2001)

This reverts commit a59a38497a29115f2fc93a92bf0bea5070869bf4.

plemarquand committed:
Ignore `--scratch-path` when calling `swift package plugin --list` (#1996)

matthewbastien committed:
combine the two fixed sections in the changelog into one (#2002)

dependabot[bot] committed:
Bump the all-dependencies group with 6 updates (#2003)

Bumps the all-dependencies group with 6 updates:

| Package | From | To |
| --- | --- | --- |
| [@typescript-eslint/eslint-plugin](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/eslint-plugin) | `8.48.0` | `8.49.0` |
| [@typescript-eslint/parser](https://github.com/typescript-eslint/typescript-eslint/tree/HEAD/packages/parser) | `8.48.0` | `8.50.0` |
| [esbuild](https://github.com/evanw/esbuild) | `0.27.0` | `0.27.1` |
| [prettier](https://github.com/prettier/prettier) | `3.7.3` | `3.7.4` |
| [replace-in-file](https://github.com/adamreisnz/replace-in-file) | `8.3.0` | `8.4.0` |
| [winston](https://github.com/winstonjs/winston) | `3.18.3` | `3.19.0` |


Updates `@typescript-eslint/eslint-plugin` from 8.48.0 to 8.49.0
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v8.49.0/packages/eslint-plugin)

Updates `@typescript-eslint/parser` from 8.48.0 to 8.50.0
- [Release notes](https://github.com/typescript-eslint/typescript-eslint/releases)
- [Changelog](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/parser/CHANGELOG.md)
- [Commits](https://github.com/typescript-eslint/typescript-eslint/commits/v8.50.0/packages/parser)

Updates `esbuild` from 0.27.0 to 0.27.1
- [Release notes](https://github.com/evanw/esbuild/releases)
- [Changelog](https://github.com/evanw/esbuild/blob/main/CHANGELOG.md)
- [Commits](https://github.com/evanw/esbuild/compare/v0.27.0...v0.27.1)

Updates `prettier` from 3.7.3 to 3.7.4
- [Release notes](https://github.com/prettier/prettier/releases)
- [Changelog](https://github.com/prettier/prettier/blob/main/CHANGELOG.md)
- [Commits](https://github.com/prettier/prettier/compare/3.7.3...3.7.4)

Updates `replace-in-file` from 8.3.0 to 8.4.0
- [Release notes](https://github.com/adamreisnz/replace-in-file/releases)
- [Changelog](https://github.com/adamreisnz/replace-in-file/blob/main/CHANGELOG.md)
- [Commits](https://github.com/adamreisnz/replace-in-file/commits)

Updates `winston` from 3.18.3 to 3.19.0
- [Release notes](https://github.com/winstonjs/winston/releases)
- [Changelog](https://github.com/winstonjs/winston/blob/master/CHANGELOG.md)
- [Commits](https://github.com/winstonjs/winston/compare/v3.18.3...v3.19.0)

---
updated-dependencies:
- dependency-name: "@typescript-eslint/eslint-plugin"
  dependency-version: 8.49.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
  dependency-group: all-dependencies
- dependency-name: "@typescript-eslint/parser"
  dependency-version: 8.50.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
  dependency-group: all-dependencies
- dependency-name: esbuild
  dependency-version: 0.27.1
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: all-dependencies
- dependency-name: prettier
  dependency-version: 3.7.4
  dependency-type: direct:development
  update-type: version-update:semver-patch
  dependency-group: all-dependencies
- dependency-name: replace-in-file
  dependency-version: 8.4.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
  dependency-group: all-dependencies
- dependency-name: winston
  dependency-version: 3.19.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
  dependency-group: all-dependencies
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

matthewbastien committed:
pin VS Code version for testing to 1.106.3 (#2010)

matthewbastien committed:
update changelog for 2.14.3 (#2013)

matthewbastien committed:
Update to the latest version of `swift-docc-render` (#1964)

matthewbastien committed:
Add fallback in toolchain selection for Swiftly.inUseVersion() when using swiftly 1.0.1 (#2014)

matthewbastien committed:
bump NodeJS version in package.json and .nvmrc (#2015)

rockbruno committed:
Fix SwiftPM's dropdown showing on projects disabling the SwiftPM integration (#2011)

matthewbastien committed:
Disallow asynchronous import statements (#2017)

amanmaurya92 committed:
Merge branch 'main' into fix/breakpoint-verified-fallback

amanmaurya92 committed:
Merge branch 'main' into fix/breakpoint-verified-fallback

matthewbastien commented:
Closing for inactivity.
